DO NOT MERGE fixes #6226 - if only one option for required field on new host page (ex. installation media), then automatically select it

isratrade opened the pull request.

Commit:
fixes #6226 - if only one option for required field on new host page (ex. installation media), then automatically select it

domcleal commented:
I've got to admit, I quite liked the change in behaviour when blank_or_inherit_f was introduced.  The main reason was for host _edit_ rather than for new hosts, as people often found it confusing when they'd edit a host to look at which attributes were currently set and it would select the only option available - even when the attribute was actually nil.

Maybe the behaviour should be different based on `new_record?`, so it's automatically selected for new hosts, but when editing you can see nil fields.

(edit: taxonomy and puppet master fields were a common problem, think the taxonomy ones still are)

ohadlevy commented:
:+1: for new_record

isratrade commented:
@domcleal, will add restriction to `new_record?` but the the original behavior from `blank_or_inherit_f` only affects `hostgroups`, not `hosts`.  This PR **only** affects `hosts`

isratrade commented on app/helpers/application_helper.rb:
```diff
@@ -366,11 +366,15 @@ def readonly_field(object, property, options={})
+      inherited_value ||= _("no value")
+      _("Inherit parent (%s)") % inherited_value
+    else
+      return true if f.object.new_record? || collection.nil?
```
rebased after adding check `if f.object.new_record?`

isratrade commented:
[test]

domcleal commented:
The bulk of blank_or_inherit_f did indeed affect host groups only, but for hosts it would always return 'true' so for hosts it always ensured there was a blank entry.

The behaviour in the current commit doesn't look like it solves the problem I mentioned.  Shouldn't it be `!f.object.new_record?` so when editing a host you always get a blank entry, and the conditional for collection.size is only reached on new hosts?

The PR's "do not merge" too - why, and what's the plan?  Does it need extending to other fields?

Commit:
fixes #6400 - show * next to required fields on forms

Commit:
fixes #6401 - change include blank option to 'No environments' or 'No subnets', etc if there are no select options
              change include blank option to 'Select environment' or 'Select subnet' if required field and more than one option

Commit:
fixes #6432 - attributes missing :presence => true

isratrade commented:
I'm not sure what in this PR would cause the integration failure on realm edit page that I fixed on PR #1547.

isratrade commented on test/integration/realm_test.rb:
```diff
@@ -9,6 +9,7 @@ class RealmTest < ActionDispatch::IntegrationTest
   test "create new page" do
     assert_new_button(realms_path,"New Realm",new_realm_path)
     fill_in "realm_name", :with => "EXAMPLE.COM"
+    select "FreeIPA", :from => "realm_realm_type"
```
I fixed the realm test issue, but PR #1547 won't hurt to merge. I still don't know why it gave my a routing error for 'myrealm.net' when the issue was with the validation.

isratrade replied:
I figured it out. The `No route matches [GET] "/realms/myrealm.net/edit"` was on the friendly_id branch

domcleal commented on app/helpers/application_helper.rb:
```diff
@@ -366,11 +366,28 @@ def readonly_field(object, property, options={})
+    if cnt == 1 && is_required?(f, attr)
+      false # auto-select option
+    elsif cnt > 0 && is_required?(f, attr)
+      blank_msg || (_("Select %s") % attr.to_s.humanize.downcase)
```
Using humanize on model names won't translate them.. this is tricky, it may depend on #4503.

ditto below

isratrade replied:
I could use .gsub('_id', '') instead of .humanize, but it will do the nearly the same. I liked .humanize since it adds a space 

```
 > 'smart_proxy_id'.humanize
 => "Smart proxy" 
```

verses

```
> 'smart_proxy_id'.gsub('_id','')
 => "smart_proxy"
```

It's just to generate a string for the 'blank message'. Why does this affect translations?

domcleal replied:
The problem is that the noun doesn't get translated as it is... you want to be able to say "Choisir un environnement" rather than "Choisir un environment", and it touches upon an issue that's seen in the UI in a few other places, that concatenating strings like this isn't i18n-friendly (e.g. in that example, the gender of the noun varies, so doesn't make sense to translate "Select %s" to "Select un".)

Really I think these strings would need to be generated in full into locale/model_attributes.rb or similar so they can be translated in their entirety, instead of using humanize and concatenation.

domcleal commented on app/helpers/layout_helper.rb:
```diff
@@ -149,10 +150,11 @@ def field(f, attr, options = {})
       content_tag :div, :class => "form-group #{error.empty? ? "" : 'has-error'}",
                   :id          => options.delete(:control_group_id) do
 
+        required_mark = '* '.html_safe if is_required?(f, attr)
```
I wonder about putting this on the right hand side of the box and making it red (think they usually are)?  Unsure..

isratrade replied:
@ohadlevy, your thoughts on this? either way is fine with me.

domcleal commented on app/helpers/application_helper.rb:
```diff
@@ -366,11 +366,28 @@ def readonly_field(object, property, options={})
+  end
+
+  def include_blank_value(f, attr, collection = nil, blank_msg = nil, none_msg = nil)
+    cnt = collection.try(:count)
```
.size preferred?

isratrade replied:
.size vs .length vs .count, I remember there are slight differences, but not totally sure. I can change to .size

domcleal replied:
.size is the more intelligent one IIRC

iNecas replied:
Also, could we use `size` or `count` instead of `cnt` variable name?

domcleal commented:
I've not tested much with single-value options, but I think the idea is good, will do more soon.  Talking of tests, some small tests of those application helpers may be useful and quite easy.

There are a couple of UI issues mentioned - one might be hard (i18n), but has come up in a few places, and the other may be worth asking Kyle or someone about, if that's easy.

isratrade commented:
good idea, I will write Kyle

iNecas commented on app/helpers/application_helper.rb:
```diff
@@ -366,11 +366,28 @@ def readonly_field(object, property, options={})
+    cnt = collection.try(:count)
+    if cnt == 1 && is_required?(f, attr)
+      false # auto-select option
+    elsif cnt > 0 && is_required?(f, attr)
```
if `cnt` is blank (for example when the collection is nil), you will get "undefined method `>' for nil:NilClass`", as I did when testing this PR

``` ruby
cnt = collection.try(:count) || 0
```

seemed to do the trick

iNecas commented:
It doesn't seem to work for domain -> subnet association: my test case: I have two domain -> subnet pairs, each associated with different location, when I change the location, I see the domain preselected, but in the subnets, I see "No Subnets", because the list of available subnets is not reflected by the fact, that the domain was preselected (and because I have no other domain to choose from, I can't even force to update the form with ajax to get the list of subnets after that).

I would expect the pre-selection happening on model, not view level, to avoid this kind of logical issues.

isratrade commented:
@iNecas, did this happen on the New Host form after a domain was already selected and then you updated the location. I do recall a known ajax bug about this independent of this PR.

isratrade commented:
@iNecas, can you confirm if this ajax issues between domain/subnets also happens on the 'develop' branch?

iNecas commented:
I've rebased against latest develop, I still see the same issue. Also, another good reason for not having this logic in the views is that the API would also deserve the same behavior: when I provision the host via hammer, I would expect specifying the org/location would determine the networks if possible.

isratrade commented:
@iNecas, I entered a separate issues for the ajax bugs after location/organization is changed.
http://projects.theforeman.org/issues/6814

iNecas commented:
I'm afraid merging this PR makes http://projects.theforeman.org/issues/6814 quite serious issue (as there is not possibility to change the domain to fix the ajax loading), therefore it would be nice to address these two together.

Are there any plans to do something similar to the API?

isratrade commented:
@iNecas, we will hold off on merging this until issue #6814 is fixed.

isratrade commented:
closing. will re-open when issue #6814 is fixed.
